allennlp/models/ensemble/bidaf_ensemble.py

```diff
@@ -0,0 +1,109 @@
+from overrides import overrides
+import torch
+from typing import Dict, List, Any
+
+from allennlp.models.ensemble.ensemble import Ensemble
+from allennlp.models.archival import load_archive
+from allennlp.models import Model
+from allennlp.models.reading_comprehension import BidirectionalAttentionFlow
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+from allennlp.training.metrics import SquadEmAndF1
+
+@Model.register("bidaf-ensemble")
+class BidafEnsemble(Ensemble):
+
+    def __init__(self,
+                 submodels: List[BidirectionalAttentionFlow]) -> None:
+        super(BidafEnsemble, self).__init__(submodels)
+
+        # Setting this propagates calls to .eval() so dropout is disabled on the submodels in evaluation
+        # and prediction.
+        self.linears = torch.nn.ModuleList(submodels)
+
+        self.submodels = submodels
+
+        self._squad_metrics = SquadEmAndF1()
+
+    @overrides
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                passage: Dict[str, torch.LongTensor],
+                span_start: torch.IntTensor = None,
+                span_end: torch.IntTensor = None,
+                metadata: List[Dict[str, Any]] = None) -> Dict[str, torch.Tensor]:
+
+        subresults = []
+        for i, submodel in enumerate(self.submodels):
+            subresults.append(submodel.forward(question, passage, span_start, span_end, metadata))
+
+        batch_size = len(subresults[0]["best_span"])
+
+        output = {
+                "best_span": torch.zeros(batch_size, 2)
+        }
+        for batch in range(batch_size):
+            max_vote = 0
+            span_votes = {}
+            for i, subresult in enumerate(subresults):
+                key = (subresult["best_span"].data[batch][0], subresult["best_span"].data[batch][1])
+                new_value = span_votes.get(key, []) + [i]
+                span_votes[key] = new_value
+                if len(new_value) > max_vote:
+                    max_vote = len(new_value)
+
+            # Choose the majority-vote span.
+            # If there is a tie, break it with the average confidence (span_start_probs + span_end_probs).
+            best = 0
+            max_average_confidence = 0
+            for (start, end), indices in span_votes.items():
+                votes = len(indices)
+                if votes == max_vote:
+                    average_confidence = 0
+                    for i in indices:
+                        subresult = subresults[i]
+                        average = (subresult["span_start_probs"].data[batch][start] + subresult["span_end_probs"].data[batch][end]) / 2.0
+                        if average > average_confidence:
+                            average_confidence = average
+                    if average_confidence > max_average_confidence:
+                        max_average_confidence = average_confidence
+                        best = i
```

Comment:
this whole section feels way too complex at a high level you should be able to

* collect votes
* sum up confidences
* sort by a combination of the two

my gut tells me it should only take a few lines of code

Reply:
I likely can.  This is not ready for PR or merging.

Reply:
fair enough 😀